cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -594,83 +577,122 @@ forgeTracer
+      \fts -> (fts { ftsBlocksForgedNum = ftsBlocksForgedNum fts + 1 },
+               ftsBlocksForgedNum fts + 1)
+    sum <$> threadStatsProjection fStats ftsBlocksForgedNum
+    >>= traceCounter "blocksForgedNum" tr
+  Consensus.TraceForgedInvalidBlock slot _ _ -> do
+    void $ mapForgingCurrentThreadStats slot fStats $
+      \fts -> (fts { ftsBlocksForgedInvalidNum = ftsBlocksForgedInvalidNum fts + 1 },
+               ftsBlocksForgedInvalidNum fts + 1)
+    sum <$> threadStatsProjection fStats ftsBlocksForgedInvalidNum
+    >>= traceCounter "blocksForgedInvalidNum" tr
+  Consensus.TraceNodeNotLeader slot ->
+    mapForgingCurrentThreadStats slot fStats (updateMissedSlotsWhileNotLeader slot)
+    >>= _traceTotalMissedSlots
   _ -> pure ()
+ where
+   updateMissedSlotsWhileNotLeader ::
+        SlotNo
+     -> ForgeThreadStats
+     -> (ForgeThreadStats,
+         Maybe (Word64, Word64, Word64, Word64, Word64, Word64, Word64, Word64))
+   updateMissedSlotsWhileNotLeader curSlot fts =
+     (fts { ftsNodeNotLeaderNum = notLeader
+          , ftsSlotsMissedNum = missed },
+      (if missed /= ftsSlotsMissedNum fts -- Did we have any fresh misses?
+       then Just (slotsPassed, leader, notLeader, leaderChecksIncomplete, forged, ledButDidntForge, forgedInvalid, missed)
+       else Nothing))
+    where
+      slotsPassed, notLeader, leader, forged, leaderChecksIncomplete, ledButDidntForge, forgedInvalid, missed :: Word64
+      -- Missed slots fall in four categories:
+      --   1. Leadership check didn't even initiate for a given slot,
+      --   2. Leadership check was initiated, but didn't
+      --      lead (pun intended) to a conclusion (leader/not leader),
+      --   3. Leadership was detected, but we didn't forge a block,
+      --   4. Block forged, but was found to be invalid.
+      --
+      -- Forging without adoption, however, is usually normal protocol operation.
+      missed  = leaderChecksIncomplete + ledButDidntForge + forgedInvalid
+      leaderChecksIncomplete = slotsPassed - leader - notLeader
+      leader                 = ftsNodeIsLeaderNum fts
+      notLeader              = ftsNodeNotLeaderNum fts + 1
+      ledButDidntForge       = leader - forged
+      forged                 = ftsBlocksForgedNum fts
+      forgedInvalid          = ftsBlocksForgedInvalidNum fts
+      -- The 1 is because we're completing this slot checks,
+      -- so the notLeader will have 1 as well.
+      slotsPassed            = 1 + (unSlotNo $ curSlot - ftsFirstSlot fts)
+
+   _traceTotalMissedSlots :: Maybe (Word64, Word64, Word64, Word64, Word64, Word64, Word64, Word64) -> IO ()
+   _traceTotalMissedSlots = \case
+     Nothing -> pure ()
+     Just _  -> sum <$> threadStatsProjection fStats ftsSlotsMissedNum
+                >>= traceCounter "slotsMissedNum" tr
+
+   _traceThreadCounters :: Maybe (Word64, Word64, Word64, Word64, Word64, Word64, Word64, Word64) -> IO ()
+   _traceThreadCounters = \case
```

Comment:
Why the `_`-prefix? Ah, I just realised that you swap between the two. Maybe add a comment?

Reply:
Will remove this, it's only been for temporary debugging.